torch/csrc/distributed/c10d/intra_node_comm.cpp

```diff
@@ -0,0 +1,628 @@
+  writeFn();
+  barrierCnt_ += 1;
+  if (barrierCnt_ != worldSize_) {
+    pthread_cond_wait(&cond_, &mutex_);
+  } else {
+    pthread_cond_broadcast(&cond_);
+  }
+  pthread_mutex_unlock(&mutex_);
+
+  // Phase 2: gather
+  pthread_mutex_lock(&mutex_);
+  gatherFn();
+  barrierCnt_ -= 1;
+  if (barrierCnt_ != 0) {
+    pthread_cond_wait(&cond_, &mutex_);
+  } else {
+    pthread_cond_broadcast(&cond_);
+  }
+  pthread_mutex_unlock(&mutex_);
+}
+
+static std::string generateUUID() {
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<> dis(0, 15);
+  std::uniform_int_distribution<> dis2(8, 11);
+
+  std::stringstream ss;
+  int i;
+
+  ss << std::hex;
+  for (i = 0; i < 8; i++) {
+    ss << dis(gen);
+  }
+  ss << "-";
+  for (i = 0; i < 4; i++) {
+    ss << dis(gen);
+  }
+  ss << "-4"; // Version 4 : Random
+  for (i = 0; i < 3; i++) {
+    ss << dis(gen);
+  }
+  ss << "-";
+  ss << dis2(gen);
+
+  for (i = 0; i < 3; i++) {
+    ss << dis(gen);
+  }
+  ss << "-";
+  for (i = 0; i < 12; i++) {
+    ss << dis(gen);
+  }
+
+  return ss.str();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Topology Detection
+////////////////////////////////////////////////////////////////////////////////
+
+// TODO: determine via GPU model
+static constexpr size_t kMaxNvLinks = 20;
+
+static std::ostream& operator<<(std::ostream& os, const NvlMesh& nvlMesh) {
+  std::ostringstream oss;
+  for (size_t i = 0; i < kMaxDevices; ++i) {
+    for (size_t j = 0; j < kMaxDevices; ++j) {
+      oss << nvlMesh[i][j] << " ";
+    }
+    oss << std::endl;
+  }
+  os << oss.str();
+  return os;
+}
+
+/**
+ * Query the nvlink connection among devices.
+ */
+static NvlMesh getNvlMesh(size_t worldSize) {
+  using namespace c10::cuda;
+
+  NvlMesh nvlMesh = {};
+#ifndef USE_ROCM
+  auto driverApi = DriverAPI::get();
+  if (driverApi == nullptr) {
+    return nvlMesh;
+  }
+
+  std::vector<nvmlDevice_t> devices(worldSize, 0);
+  std::unordered_map<std::string, size_t> busIdToIdx;
+  std::vector<size_t> switchLinkCount(worldSize, 0);
+
+  for (size_t idx = 0; idx < worldSize; ++idx) {
+    cudaDeviceProp prop{};
+    if (cudaGetDeviceProperties(&prop, idx) == cudaErrorInvalidDevice) {
+      continue;
+    }
+    char busId[80];
+    snprintf(
+        busId,
+        sizeof(busId),
+        NVML_DEVICE_PCI_BUS_ID_FMT,
+        prop.pciDomainID,
+        prop.pciBusID,
+        prop.pciDeviceID);
+    TORCH_CHECK(
+        driverApi->nvmlDeviceGetHandleByPciBusId_v2_(busId, &devices[idx]) ==
+        NVML_SUCCESS);
+    busIdToIdx.emplace(std::make_pair(busId, idx));
+  }
+
+  for (size_t idx = 0; idx < worldSize; ++idx) {
+    for (size_t link = 0; link < kMaxNvLinks; ++link) {
+      nvmlReturn_t ret;
+      nvmlIntNvLinkDeviceType_t deviceType;
+      ret = driverApi->nvmlDeviceGetNvLinkRemoteDeviceType_(
+          devices[idx], link, &deviceType);
+      if (ret != NVML_SUCCESS) {
+        break;
```

Comment:
Wouldn't it be more advisable to just return an empty NvlMesh from this function in case of any unexpected error codes? Otherwise we might get inconsistent NvlMeshes across ranks and that might lead to issues?